utils/profiling.py

```diff
@@ -0,0 +1,120 @@
+import cProfile
+import functools
+import io
+import os
+import pstats
+import sys
+import threading
+import time
+
+__all__ = [
+    "profiling_support",
+    "run_with_profiler",
+    "profile_call",
+    "ProfilingMiddleware",
+]
+
+
+def profiling_support(func):
+    """
+    run a function with profiler enabled, if BERTH_PROFILING_ENABLED is set to 1 in environment.
+    """
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        if get_profile_flag():
+            return profile_call(func, *args, **kwargs)
+        else:
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+def run_with_profiler(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        return profile_call(func, *args, **kwargs)
+
+    return wrapper
+
+
+def profile_call(func, *args, **kwargs):
+    profile = cProfile.Profile()
+    try:
+        profile.enable()
+        return func(*args, **kwargs)
+    finally:
+        profile.disable()
+        dump_profiling_results(
+            profile, f"{func.__name__}-{threading.current_thread().ident}"
+        )
+
+
+def dump_profiling_results(profile, identifier=None):
+    s = io.StringIO()
+    num_results = int(os.getenv("PROFILE_RESULTS", "100"))
+    sortby = os.getenv("PROFILE_SORT", "cumulative")
+    for sort_val in sortby.split(","):
+        ps = pstats.Stats(profile, stream=s).sort_stats(sort_val)
+        ps.print_stats(num_results)
+
+        if os.getenv("PROFILE_CALLEES") == "1":
+            ps.print_callees(num_results)
+
+        if os.getenv("PROFILE_CALLERS") == "1":
+            ps.print_callers(num_results)
+
+    sys.stderr.write(s.getvalue())
```

Comment:
Instead of directly printing to `stderr`, we could use [python logging](https://docs.python.org/3/library/logging.html) and set up a logger on the settings specific for the profiling.

Then, we can easily configure whether the results are printed to the console or to a file or both.